File: services/dameng-service/internal/catalog/list.go
// Package catalog 提供 SQL 补全用目录检索（与 tree 导航分离，见 docs/23 / docs/28）。
package catalog

import (
	"context"
	"database/sql"
	"fmt"
	"strings"

	"niuma/services/dameng-service/internal/meta"
	"niuma/services/dameng-service/internal/tree"
)

const (
	// DefaultLimit 是补全默认条数。
	DefaultLimit = 200
	// MaxLimit 是补全上限。
	MaxLimit = 500
)

// ListParams 是目录检索参数。
type ListParams struct {
	Prefix        string
	Limit         int
	ExcludeSystem bool
	Schema        string
	Table         string
	Types         []string
}

// SchemaHit / TableHit / ColumnHit 为补全命中。
type SchemaHit struct {
	Name string `json:"name"`
}
type TableHit struct {
	Name   string `json:"name"`
	Type   string `json:"type,omitempty"`
	Schema string `json:"schema,omitempty"`
}
type ColumnHit struct {
	Name     string `json:"name"`
	DataType string `json:"dataType,omitempty"`
	Schema   string `json:"schema,omitempty"`
	Table    string `json:"table,omitempty"`
}

// SchemasResult / TablesResult / ColumnsResult 带截断标记。
type SchemasResult struct {
	Schemas   []SchemaHit `json:"schemas"`
	Truncated bool        `json:"truncated,omitempty"`
}
type TablesResult struct {
	Tables    []TableHit `json:"tables"`
	Truncated bool       `json:"truncated,omitempty"`
}
type ColumnsResult struct {
	Columns   []ColumnHit `json:"columns"`
	Truncated bool        `json:"truncated,omitempty"`
}

func normalizeLimit(limit int) int {
	if limit <= 0 {
		return DefaultLimit
	}
	if limit > MaxLimit {
		return MaxLimit
	}
	return limit
}

// ListSchemas 列出用户/模式（ALL_USERS，可排除系统用户）。
func ListSchemas(ctx context.Context, db *sql.DB, params ListParams) (SchemasResult, error) {
	res, err := tree.ListSchemas(ctx, db, tree.ListParams{
		Filter:        params.Prefix,
		Limit:         normalizeLimit(params.Limit),
		ExcludeSystem: params.ExcludeSystem,
	})
	if err != nil {
		return SchemasResult{}, err
	}
	out := SchemasResult{Truncated: res.Truncated, Schemas: make([]SchemaHit, 0, len(res.Schemas))}
	for _, s := range res.Schemas {
		out.Schemas = append(out.Schemas, SchemaHit{Name: s.Name})
	}
	return out, nil
}

// ListTables 按 schema + 前缀检索表/视图。
func ListTables(ctx context.Context, db *sql.DB, params ListParams) (TablesResult, error) {
	schema := strings.TrimSpace(params.Schema)
	if schema == "" {
		return TablesResult{}, fmt.Errorf("dameng: catalog: schema required")
	}
	types := params.Types
	if len(types) == 0 {
		types = []string{"table", "view"}
	}
	res, err := tree.ListTables(ctx, db, tree.ListParams{
		Schema: schema,
		Filter: params.Prefix,
		Limit:  normalizeLimit(params.Limit),
		Types:  types,
	})
	if err != nil {
		return TablesResult{}, err
	}
	out := TablesResult{Truncated: res.Truncated, Tables: make([]TableHit, 0, len(res.Tables))}
	for _, t := range res.Tables {
		out.Tables = append(out.Tables, TableHit{Name: t.Name, Type: t.Type, Schema: schema})
	}
	return out, nil
}

// ListColumns 列出列（委托 meta，再按 prefix/limit 截断）。
func ListColumns(ctx context.Context, db *sql.DB, params ListParams) (ColumnsResult, error) {
	table := strings.TrimSpace(params.Table)
	if table == "" {
		return ColumnsResult{}, fmt.Errorf("dameng: catalog: table required")
	}
	schema := strings.TrimSpace(params.Schema)
	if schema == "" {
		return ColumnsResult{}, fmt.Errorf("dameng: catalog: schema required")
	}
	cols, err := meta.ListColumns(ctx, db, meta.RelationRef{Schema: schema, Name: table})
	if err != nil {
		return ColumnsResult{}, err
	}
	limit := normalizeLimit(params.Limit)
	prefix := strings.ToLower(strings.TrimSpace(params.Prefix))
	out := ColumnsResult{Columns: make([]ColumnHit, 0, len(cols.Columns))}
	for _, c := range cols.Columns {
		if prefix != "" && !strings.HasPrefix(strings.ToLower(c.Name), prefix) {
			continue
		}
		if len(out.Columns) >= limit {
			out.Truncated = true
			break
		}
		out.Columns = append(out.Columns, ColumnHit{
			Name:     c.Name,
			DataType: c.DataType,
			Schema:   schema,
			Table:    table,
		})
	}
	return out, nil
}

File: services/oracle-service/src/catalog/list.cpp
#include "catalog/list.hpp"

#include "meta/system_users.hpp"
#include "session/sql_rows.hpp"
#include "util/ident.hpp"
#include "util/sql_literal.hpp"

#include <cctype>

namespace niuma::oracle::catalog {

ListParams ListParams::FromJson(const nlohmann::json& j) {
  ListParams p;
  if (j.contains("schema") && j["schema"].is_string()) {
    p.schema = j["schema"].get<std::string>();
  }
  if (j.contains("table") && j["table"].is_string()) {
    p.table = j["table"].get<std::string>();
  }
  if (j.contains("prefix") && j["prefix"].is_string()) {
    p.prefix = j["prefix"].get<std::string>();
  }
  if (j.contains("limit") && j["limit"].is_number_integer()) {
    p.limit = j["limit"].get<int>();
  }
  return p;
}

nlohmann::json ListSchemas(session::Session& session, const ListParams& params, std::string& error) {
  const int limit = util::ClampListLimit(params.limit, 200, 2000);
  const std::string like = util::QuoteLiteral(util::LikePrefixPattern(params.prefix));
  const std::string sql =
      "SELECT USERNAME FROM ALL_USERS WHERE UPPER(USERNAME) LIKE UPPER(" + like +
      ") ESCAPE '\\' ORDER BY USERNAME";

  session::SqlRowsResult rows;
  if (!session::ExecStringRows(session, sql, limit + 1, rows, error)) {
    error.clear();
    const std::string fallback =
        "SELECT USERNAME FROM USER_USERS WHERE UPPER(USERNAME) LIKE UPPER(" + like +
        ") ESCAPE '\\' ORDER BY USERNAME";
    if (!session::ExecStringRows(session, fallback, limit + 1, rows, error)) {
      return {};
    }
  }

  nlohmann::json schemas = nlohmann::json::array();
  bool truncated = rows.truncated;
  for (const auto& row : rows.rows) {
    if (row.empty()) {
      continue;
    }
    if (meta::IsSystemSchema(row[0])) {
      continue;
    }
    if (static_cast<int>(schemas.size()) >= limit) {
      truncated = true;
      break;
    }
    schemas.push_back({{"name", row[0]}});
  }
  nlohmann::json out{{"schemas", schemas}};
  if (truncated) {
    out["truncated"] = true;
  }
  return out;
}

nlohmann::json ListTables(session::Session& session, const ListParams& params, std::string& error) {
  if (params.schema.empty() || !util::IsSafeIdent(params.schema)) {
    error = "oracle: schema required";
    return {};
  }
  const int limit = util::ClampListLimit(params.limit, 200, 2000);
  const std::string owner = util::QuoteLiteral(params.schema);
  const std::string like = util::QuoteLiteral(util::LikePrefixPattern(params.prefix));
  const std::string sql =
      "SELECT OBJECT_NAME, OBJECT_TYPE FROM ALL_OBJECTS WHERE OWNER = " + owner +
      " AND OBJECT_TYPE IN ('TABLE','VIEW') AND OBJECT_NAME LIKE " + like +
      " ESCAPE '\\' ORDER BY OBJECT_NAME";

  session::SqlRowsResult rows;
  if (!session::ExecStringRows(session, sql, limit + 1, rows, error)) {
    return {};
  }

  nlohmann::json tables = nlohmann::json::array();
  bool truncated = rows.truncated;
  for (const auto& row : rows.rows) {
    if (row.size() < 2) {
      continue;
    }
    if (static_cast<int>(tables.size()) >= limit) {
      truncated = true;
      break;
    }
    std::string typ = row[1];
    for (char& c : typ) {
      c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    }
    tables.push_back({{"name", row[0]}, {"type", typ}, {"schema", params.schema}});
  }
  nlohmann::json out{{"tables", tables}};
  if (truncated) {
    out["truncated"] = true;
  }
  return out;
}

nlohmann::json ListColumns(session::Session& session, const ListParams& params, std::string& error) {
  if (params.schema.empty() || !util::IsSafeIdent(params.schema)) {
    error = "oracle: schema required";
    return {};
  }
  if (params.table.empty() || !util::IsSafeIdent(params.table)) {
    error = "oracle: table required";
    return {};
  }
  const int limit = util::ClampListLimit(params.limit, 200, 2000);
  const std::string owner = util::QuoteLiteral(params.schema);
  const std::string table = util::QuoteLiteral(params.table);
  const std::string like = util::QuoteLiteral(util::LikePrefixPattern(params.prefix));
  const std::string sql =
      "SELECT COLUMN_NAME, DATA_TYPE FROM ALL_TAB_COLUMNS WHERE OWNER = " + owner +
      " AND TABLE_NAME = " + table + " AND COLUMN_NAME LIKE " + like +
      " ESCAPE '\\' ORDER BY COLUMN_ID";

  session::SqlRowsResult rows;
  if (!session::ExecStringRows(session, sql, limit + 1, rows, error)) {
    return {};
  }

  nlohmann::json columns = nlohmann::json::array();
  bool truncated = rows.truncated;
  for (const auto& row : rows.rows) {
    if (row.empty()) {
      continue;
    }
    if (static_cast<int>(columns.size()) >= limit) {
      truncated = true;
      break;
    }
    nlohmann::json col{{"name", row[0]}, {"schema", params.schema}, {"table", params.table}};
    if (row.size() > 1) {
      col["dataType"] = row[1];
    }
    columns.push_back(std::move(col));
  }
  nlohmann::json out{{"columns", columns}};
  if (truncated) {
    out["truncated"] = true;
  }
  return out;
}

}  // namespace niuma::oracle::catalog
